### PR TITLE
Define guideline for components to deprecate config options

### LINKED ISSUE #5686: Define guideline for components to deprecate configuration
The following [change](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/10821) in contrib caused a breaking change for a component identified as `stable`. This would cause existing users of the component to break when upgrading their collector if they were using the configuration options removed.

Currently the [policy](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#breaking-changes) for breaking changes is defined for changes to the API, but it's unclear how components should apply this policy for configuration changes.

This issue is a discussion to determine:
1. what the guidelines should be for configuration changes/deprecation
2. how components can be marked stable and if they even can be stable while the collector is still in beta
3. if there should be separate versioning for individual components. Currently all components within a release have the same version. Separating component versions could allow components to change major versions and follow semver, although it's still unclear if this would be enough considering the overall version of the collector wouldn't change.

Pinging the folks that were involved in a discussion in a private channel to add their comments here @Aneurysm9 @jpkrohling @dashpole @mx-psi 

_Note: I understand the issue was caused by a change in the contrib repo, but since the stability definition is in this repo, I figured it made more sense to open the issue here._

### PR DESCRIPTION
**Pull request by jpkrohling**
Fixes #5686

Signed-off-by: Juraci Paixão Kröhling <juraci@kroehling.de>

### CODE REVIEW

**Review by dashpole — APPROVED** [review]

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -494,6 +494,51 @@ that each of the following steps is done in a separate version:
...
+latest version of the component.
+
+When adding a new required option:
+- the option MUST come with a sensible default value
+
+When renaming or removing a configuration option:
+- the option MUST be deprecated in one version
+- a WARN level message should be logged, with a link to a place within the component's repository where the change is
+  documented and a migration path is provided
+- the option MUST be kept for at least N+1 version, and MAY be hidden behind a feature flag in N+2
```
Comment (codeboten) [inline CONTRIBUTING.md:529]:
```suggestion
- the option MUST be kept for at least N+1 version, and MAY be hidden behind a feature gate in N+2
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -494,6 +494,51 @@ that each of the following steps is done in a separate version:
...
+
+Additionally, when removing an option:
+- the option MAY be made non operational already by the same version where it is deprecated
+
+##### Stable components
+
+Stable component MUST be compatible between minor versions, unless critical security issues are found. In that case, the
+component owner MUST provide a migration path and a reasonable time frame for users to upgrade. The same rules from beta
+components apply to stable when it comes to configuration changes.
+
```
Comment (codeboten) [inline CONTRIBUTING.md:540]:
```suggestion
```

**Review by codeboten — APPROVED** [review]
Looks good, just a couple of minor suggestions

**Review by mx-psi — APPROVED** [review]

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -494,6 +494,50 @@ that each of the following steps is done in a separate version:
...
    needed.
 
+#### Configuration changes
+
+##### Alpha components
+
+Configuration for alpha components can be changed with minimal notice. Documenting them as part of the changelog is
+sufficient. We still recommend giving users one or two minor version's notice before breaking the configuration, such as
+when removing or renaming a configuration option. Providing a migration path in the component's repository is NOT
+required for alpha components, altough still recommended.
```
Comment (codeboten) [inline CONTRIBUTING.md:504]:
```suggestion
required for alpha components, although still recommended.
```

**Review by codeboten — APPROVED** [review]

**Pull request merged**